src/optimizer/rules/eliminate_cse_rule.rs

```diff
@@ -0,0 +1,331 @@
+// Copyright 2022 RisingLight Project Authors. Licensed under Apache-2.0.
+
+use sqlparser::ast::BinaryOperator;
+
+use super::*;
+use crate::binder::*;
+use crate::optimizer::plan_nodes::{
+    IntoPlanRef, LogicalAggregate, LogicalProjection, PlanTreeNodeUnary,
+};
+
+pub struct ProjectEliminateCSE {}
+pub struct AggregateEliminateCSE {}
+
+impl Rule for ProjectEliminateCSE {
+    fn apply(&self, plan: PlanRef) -> Result<PlanRef, ()> {
+        let projection = plan.as_logical_projection()?;
+        let mut cse_eliminator = CSEEliminator::new();
+        let mut proj_exprs = projection.project_expressions().to_vec();
+        // Search Phase
+        cse_eliminator.search(proj_exprs.iter());
+
+        if !cse_eliminator.can_collect() {
+            // No common sub-expressions
+            return Ok(plan);
+        }
+
+        // Collect Phase
+        cse_eliminator.collect(proj_exprs.iter());
+
+        // Rewrite Phase
+        cse_eliminator.replace(proj_exprs.iter_mut());
+
+        let child =
+            LogicalProjection::new(cse_eliminator.new_projection_exprs(), projection.child());
+        Ok(LogicalProjection::new(proj_exprs, child.into_plan_ref()).into_plan_ref())
+    }
+}
+
+impl Rule for AggregateEliminateCSE {
+    fn apply(&self, plan: PlanRef) -> Result<PlanRef, ()> {
+        let agg = plan.as_logical_aggregate()?;
+        let mut cse_eliminator = CSEEliminator::new();
+        let mut group_keys = agg.group_keys().to_vec();
+        let mut agg_calls = agg.agg_calls().to_vec();
+        // Search Phase
+        let exprs_iter = group_keys
+            .iter()
+            .chain(agg_calls.iter().flat_map(|calls| calls.args.iter()));
+        cse_eliminator.search(exprs_iter.clone());
+
+        if !cse_eliminator.can_collect() {
+            // No common sub-expressions
+            return Ok(plan);
+        }
+
+        // Collect Phase
+        cse_eliminator.collect(exprs_iter);
+
+        let exprs_iter = group_keys
+            .iter_mut()
+            .chain(agg_calls.iter_mut().flat_map(|calls| calls.args.iter_mut()));
+        cse_eliminator.replace(exprs_iter);
+
+        let child = LogicalProjection::new(cse_eliminator.new_projection_exprs(), agg.child());
+        Ok(LogicalAggregate::new(agg_calls, group_keys, child.into_plan_ref()).into_plan_ref())
+    }
+}
+
+#[derive(PartialEq, Debug)]
+enum EliminatorState {
+    Search,
+    Collect,
+    Replace,
+    End,
+}
```

Comment:
I think `State` seems not necessaray. What about using different structs to do different phases instead of state transition?

Reply:
> I think `State` seems not necessaray. What about using different structs to do different phases instead of state transition?

It seems that  ```expr_maps: HashMap<BoundExpr, SubExpr>``` will be accessed during the search phase and collect phase. Using diffrent structs to access and modify `expr_maps` maybe look likes wierd?  But `Replace` state and `End` state can be removed, they are just to avoid me doing something wrong.

Reply:
> It seems that expr_maps: HashMap<BoundExpr, SubExpr> will be accessed during the search phase and collect phase. Using diffrent structs to access and modify expr_maps maybe look likes wierd?

IMHO it makes sense for each phase to process some data and then pass data to the next phase. Using state machine to model the processing pipeline makes the logic of different tasks mixed together and hard to read (And I didn't fully understand it yet) :( 

What about:
```rust
/// output balabala, modify balabala
fn eliminate(exprs: &mut Vec<BoundExpr>) -> Option<Vec<BoundExpr>> {
	let expr_maps = search(exprs.iter())?;
	let child_projection_exprs = collect(exprs.iter(), expr_maps);
	rewrite(exprs, child_projection_exprs);
	Some(child_projection_exprs)
}

/// do balabala, output balabala
fn search(exprs: &Vec<BoundExpr>) -> Option<HashMap<BoundExpr, SubExpr>> {
	struct Visitor { expr_maps }
	impl ExprVisitor for Visitor {...}
	let visitor = ...;
	...
    visitor.expr_maps
}

/// do balabala
fn collect(exprs: &Vec<BoundExpr>, subexprs: ...) ->  Vec<BoundExpr> {
	struct Visitor {...}
	impl ExprVisitor for Visitor {...}
	...
}

/// do balabala
fn rewrite(exprs: &mut Vec<BoundExpr>, child_projection_exprs) {
	struct Rewriter {...}
	...
}
```